Feat upgradeable account registry

thomas-waite opened the pull request:
## Summary
This PR makes the `AccountRegistry.sol` contract, first introduced on the `feat-sdk-ui` branch upgradeable. It makes use of the unstructured proxy storage pattern developed by OpenZeppelin.

## Description
The `AccountRegistry.sol` contract is used to register register Ethereum accounts that will use the AZTEC SDK.  It needs to be upgradeable to account for the case where new methods are added to register users, but also to account for entirely new behaviour if the contract purpose evolves. 

### Architecture
The `AccountRegistry` contract has now been broken out into two contracts. There are `Behaviour` contracts, which define the methods of the behaviour and a storage contract in the form of a `Proxy` - this maintains state. The `Behaviour` contracts are the mutable, upgradeable aspect of the system.

In addition, there is now an `AccountRegistryManager` contract, which has two key methods:
- `deployProxy()`
- `upgradeAccountRegistry()` 

which are used to codify the process of deploying the proxy contract and also to upgrade the current account registry implementation. The use of this contract reduces the scope for human error in performing these actions. 

It should be noted that in the pattern implemented here, `Behaviour` contracts have to be deployed manually - there is no concept of `Factory` contracts like there is in our upgradeable note registry work. 

### Versioning system
A basic versioning system for the `Behaviour` contracts has been implemented. A basic `epoch` based system has been used, where each `Behaviour` contract is assigned an integer value. 

The system is initialised with an `epoch` of `0`, and then each `Behaviour` contract has an incremented epoch. Each behaviour contract must inherit in order from behaviour contracts with a lower epoch number, and it is only possible to upgrade to contracts with a higher epoch than the `latestEpoch` - in order to prevent any potential corruption of storage. 

Unlike upgradeable note registries, the `AccountRegistry` contracts are `assetType` and `cryptoSystem` agnostic - there is no need to account for these variables in the versioning system.

<!--- Describe your changes in detail -->

## Testing instructions
Various tests for the pattern have been added in `protocol/test/AccountRegistry`.


## Types of changes
* New feature (non-breaking change which adds functionality)
Implement upgradeable account registries

thomas-waite committed:
refactor(protocol): break Proxy contracts out into own directory + test
Proxy pattern will be reused by the AccountRegistry upgradeability pattern, so makes sense to dissociate it exclusively from the Note Registry upgradeability pattern

thomas-waite committed:
feat(protocol): add interfaces for AccountRegistry contracts

thomas-waite committed:
feat(protocol): add AccountRegistry contracts from feat-sdk-ui branch

thomas-waite committed:
feat(protocol): implement AccountRegistryManager contract

thomas-waite committed:
test(protocol): add AccountRegistry tests from feat-sdk-ui branch

thomas-waite committed:
refactor: add sig schema and helpers for account registry

thomas-waite committed:
feat: complete manager.deployProxy() functionality and test

thomas-waite committed:
refactor(protocol): replace constructors with initialize functions
done to make compatible with the upgrade pattern

thomas-waite committed:
refactor(protocol): add versioning system for registry behaviours

thomas-waite committed:
build: update yarn.lock

thomas-waite committed:
style(aztec.js): minor var rename, apply prettier and eslint

thomas-waite committed:
feat(protocol): add modifier contract with checkZeroAddress()

thomas-waite committed:
docs(protocol): add docs for behaviour contract and var renaming

thomas-waite committed:
refactor(protocol): ensure behaviour can only be initialised once

ArnSch commented on packages/protocol/contracts/AccountRegistry/AccountRegistryManager.sol:
```diff
@@ -0,0 +1,107 @@
+     * @param trustedGSNSignerAddress - address that is being trusted to produce signatures to approve 
+     * relayed GSN calls
+     */
+    function deployProxy(
```
We should never be able to deploy a new Proxy contract, so think nicest way to ensure it only gets called once is to make it the constructor.

thomas-waite replied:
Good idea, have done

ArnSch commented on packages/protocol/contracts/AccountRegistry/AccountRegistryManager.sol:
```diff
@@ -0,0 +1,107 @@
+    
+        proxy = proxyAddress;
+        behaviour = IAccountRegistryBehaviour(proxyAddress);
+        accountRegistry = initialBehaviourAddress;
```
I think these variables should be collapsed into 1 (just the proxy address).

You currently don't use `accountRegistry` anywhere, and in the upgrade flow, you pass in the `proxyAddress` (and so don't read from `proxy`). 

I also find it slightly odd if any callers to the account registry need to go through the manager contract (i.e. having the `behaviour` variable). Can we get away with simply storing the address of the proxy, and delegating the responsibility of handling the interface to any calling contract?

thomas-waite replied:
Agreed, have removed the unnecessary variables

ArnSch replied:
I don't think you've fully removed them: you still declare `accountRegistry` and `behaviour`, and set `accountRegistry` in `upgradeAccountRegistry` (think that's unnecessary given you never read from it).

thomas-waite replied:
Yep, not sure why they're still there. Have now removed

ArnSch commented on packages/protocol/contracts/AccountRegistry/AccountRegistryManager.sol:
```diff
@@ -0,0 +1,107 @@
+     * @param newBehaviourAddress - address of the behaviour contract to be upgraded to
+    */
+    function upgradeAccountRegistry(
+        address proxyAddress,
```
Linked to comment above, this shouldn't be passed in, but should read from the state of the contract.

thomas-waite replied:
Good point, done

ArnSch commented on packages/protocol/contracts/AccountRegistry/TransactionRelayer.sol:
```diff
@@ -0,0 +1,49 @@
+
+    ACEModule.ACE ace;
+
+    function initialize(address _ace) public initializer {
```
Just sense checking that it's ok for this to be callable (only once) by anyone?

thomas-waite replied:
Have made the visibility modifiers for all `initialize()` methods, of contracts called by the Behaviour contract's `initialize()` function, `internal`. This works as the `Behaviour` contract is derived from these contracts. 

It stops any external party calling `initialize()` on `TransactionRelayer()` and supplying a false `_ace` address for example.

ArnSch commented on packages/protocol/contracts/AccountRegistry/epochs/1/Behaviour1.sol:
```diff
@@ -0,0 +1,85 @@
+    * @param _aceAddress - address of the AZTEC Cryptography Engine
+    * @param _trustedGSNSignerAddress - address which will produce signature to approve relayed GSN calls
+    */
+    function initialize(address _aceAddress, address _trustedGSNSignerAddress) initializer public {
```
Sorry if I missed, but do you test this initialisation works as expected anywhere?

thomas-waite replied:
Nope, this doesn't appear to have been originally tested. 

I've added in a test that checks the initialisation is working - we check that the appropriate event is emitted.

ArnSch commented on packages/protocol/contracts/AccountRegistry/epochs/1/base/BehaviourBase1.sol:
```diff
@@ -0,0 +1,82 @@
+
+    event LogAddress(
+        address account
+    );
```
```suggestion
    );
    
```

ArnSch commented on packages/protocol/contracts/AccountRegistry/epochs/1/base/BehaviourBase1.sol:
```diff
@@ -0,0 +1,82 @@
+    * @dev function to validate the user is either the sender or has submitted an EIP712 signature from the 
+    * account they are registering in the AZTEC extension.
+    **/
+    function hashAZTECAccount(AZTECAccount memory _AZTECAccount) internal view returns (bytes32){
```
```suggestion
    function hashAZTECAccount(AZTECAccount memory _AZTECAccount) internal view returns (bytes32) {
```

thomas-waite replied:
Done

ArnSch commented on packages/protocol/contracts/AccountRegistry/epochs/1/base/BehaviourBase1.sol:
```diff
@@ -0,0 +1,82 @@
+
+
+    /**
+    * @dev function to validate the user is either the sender or has submitted an EIP712 signature from the 
```
This function doesn't seem to do any validation.

thomas-waite replied:
Looks like the docs are out of date with what the function does. Have updated

ArnSch commented on packages/protocol/contracts/AccountRegistry/epochs/1/base/BehaviourBase1.sol:
```diff
@@ -0,0 +1,82 @@
+        );
+        require(_account == signer, 'signer must be the account');
+        accountMapping[_account] = _linkedPublicKey;
+        // emit event EventService
```
?

thomas-waite replied:
Have tidied up and removed the comment

ArnSch commented on packages/protocol/contracts/AccountRegistry/epochs/1/base/BehaviourBase1.sol:
```diff
@@ -0,0 +1,82 @@
+            hashAZTECAccount(AZTECAccount(_account, _linkedPublicKey)),
+            _signature
+        );
+        require(_account == signer, 'signer must be the account');
```
Anyone can replay this signature, and emit a new arbitrary `_spendingPublicKey`. Can this be used to grief legitimate users? Why not include the _spendingPublicKey in the hash?

Would replays/superfluous events mess with UX/UI @LeilaWang @joeandrews ?

thomas-waite replied:
This is a good point, and it shouldn't be possible to replay signatures. 

Will check with @joeandrews  but also want to include the `_spendingPublicKey` in the hash

thomas-waite replied:
Have added signature replay protection - now checks that the signature hasn't been seen before. 

`_spendingPublicKey` can't be easily mixed into the hash as it's generated at a different/later stage to the signature

ArnSch commented on packages/protocol/contracts/AccountRegistry/interfaces/IAccountRegistryBehaviour.sol:
```diff
@@ -0,0 +1,37 @@
+pragma solidity >=0.5.0 <0.6.0;
+
+
```
```suggestion
```

ArnSch commented on packages/protocol/contracts/AccountRegistry/interfaces/IAccountRegistryBehaviour.sol:
```diff
@@ -0,0 +1,37 @@
+        bytes spendingPublicKey 
+    );
+
+    event LogAddress(
+        address account
+    );
+
+    event LogString(
+        string message
+    );
+
+    event LogBytes(
+        bytes32 sig
+    );
```
We could clean up the base contract by having it extend this interface. No strong opinions though.

thomas-waite replied:
Good idea, done

ArnSch commented on packages/protocol/test/AccountRegistry/AccountRegistryManager.js:
```diff
@@ -0,0 +1,233 @@
+                const expectedNewBehaviourAddress = testBehaviour.address;
+                expect(newBehaviourAddress).to.equal(expectedNewBehaviourAddress);
+
+                const isNewFeatureImplemented = 'newFeature()' in testBehaviour.methods;
```
This just tests that the contract you're using as a behaviour has the new method you put on that same contract. Would be nice to see some calls to the proxy which more explicitly test that upgrades/initialisations have happened correctly.

thomas-waite replied:
Yep agreed. Have now adapted that test to call a `newFeature()` function on the upgraded behaviour contract via the `proxyContract`

ArnSch commented on packages/protocol/test/AccountRegistry/AccountRegistryManager.js:
```diff
@@ -0,0 +1,233 @@
+                const preUpgradeProxy = await manager.proxy.call();
+                await manager.upgradeAccountRegistry(preUpgradeProxy, testBehaviour.address);
+
+                const storedLinkedPublicKey = await behaviour.accountMapping.call(address);
```
This is completely avoiding the Proxy contract, only testing that the deployed behaviour is setting state (in itself, not in proxy context) correctly.

thomas-waite replied:
Good point, have updated the tests

ArnSch commented on packages/protocol/test/AccountRegistry/AccountRegistry.js:
```diff
@@ -0,0 +1,87 @@
+            registryContract = await AccountRegistry.new({ from: accounts[0] });
+        });
+
+        it('be able to register the extension with a valid signature', async () => {
```
```suggestion
        it('should be able to register the extension with a valid signature', async () => {
```

thomas-waite replied:
Done

ArnSch reviewed (changes requested):
Some updates needed around tests, and I think we can clean up the behaviour/manager a bit.

thomas-waite committed:
refactor(protocol): implement PR feedback
move deployProxy() functionality into constructor

thomas-waite committed:
refactor(protocol): implement PR feedback
add sig replay protection to registerAZTECExtension()
update function doc string
make BehaviourBase inherit from IAccountRegistryBehaviour

thomas-waite committed:
refactor(protocol): change epoch versioning system
changed to be date based rathert than integer based

thomas-waite committed:
refactor(protocol): implement PR feedback
make initialize() functions internal
add test to check initialize() functions perform as expected

ArnSch commented on packages/protocol/contracts/AccountRegistry/AccountRegistryManager.sol:
```diff
@@ -0,0 +1,104 @@
+    using SafeMath for uint256;
+
+    IAccountRegistryBehaviour public behaviour;
+    address public accountRegistry;
```
```suggestion
```

thomas-waite replied:
removed

ArnSch commented on packages/protocol/contracts/AccountRegistry/AccountRegistryManager.sol:
```diff
@@ -0,0 +1,104 @@
+contract AccountRegistryManager is Ownable, Modifiers {
+    using SafeMath for uint256;
+
+    IAccountRegistryBehaviour public behaviour;
```
```suggestion
```

thomas-waite replied:
removed

ArnSch commented on packages/protocol/contracts/AccountRegistry/AccountRegistryManager.sol:
```diff
@@ -0,0 +1,104 @@
+
+        ProxyAdmin(proxyAddress).upgradeTo(newBehaviourAddress);
+        
+        accountRegistry = newBehaviourAddress;
```
```suggestion
```

thomas-waite replied:
removed

ArnSch commented on packages/protocol/test/AccountRegistry/AccountRegistry.js:
```diff
@@ -0,0 +1,92 @@
+        });
+    });
+
+    describe('Failure States', () => {
```
No replay protection test?

thomas-waite replied:
Sure, have added

ArnSch commented on packages/protocol/test/AccountRegistry/AccountRegistryManager.js:
```diff
@@ -0,0 +1,206 @@
+                expect(parseInt(logs[0].topics[2], 16)).to.equal(parseInt(proxyAdmin, 16));
+            });
+
+            it('should set epoch number on proxy', async () => {
```
```suggestion
            it('should set epoch number on manager', async () => {
```

thomas-waite replied:
done

ArnSch commented on packages/protocol/test/AccountRegistry/AccountRegistryManager.js:
```diff
@@ -0,0 +1,206 @@
+                const testBehaviour = await TestBehaviour.new();
+                await manager.upgradeAccountRegistry(testBehaviour.address);
+
+                proxyContract = await TestBehaviour.at(proxyAddress);
```
Minor, but why the reassignment?

thomas-waite replied:
There's no need for it, have removed

ArnSch reviewed (changes requested):
A couple of minor things, and waiting on initialisation pattern sense check.

thomas-waite committed:
refactor(protocol): implement PR feedback
remove unnecessary variables
add replay protection test

ArnSch reviewed (approved):
LGTM!
